src/Tmds.Ssh/UserAuthentication.PublicKeyAuth.cs

```diff
@@ -224,45 +224,127 @@ byte	3
+                error = null;
+                return true;
+            }
+            catch (Exception ex)
+            {
+                error = new FormatException($"The data can not be parsed into an RSA key.", ex);
+                rsa.Dispose();
+                return false;
+            }
+        }
+
+        private static bool TryParseOpenSshEcdsaKey(Name keyIdentifier, SequenceReader reader, [NotNullWhen(true)] out PrivateKey? privateKey, [NotNullWhen(false)] out Exception? error)
+        {
+            privateKey = null;
+
+            Name curveName = reader.ReadName();
+
+            HashAlgorithmName allowedHashAlgo;
+            ECCurve curve;
+            if (curveName == AlgorithmNames.Nistp256)
+            {
+                allowedHashAlgo = HashAlgorithmName.SHA256;
+                curve = ECCurve.NamedCurves.nistP256;
+            }
+            else if (curveName == AlgorithmNames.Nistp384)
+            {
+                allowedHashAlgo = HashAlgorithmName.SHA384;
+                curve = ECCurve.NamedCurves.nistP384;
+            }
+            else if (curveName == AlgorithmNames.Nistp521)
+            {
+                allowedHashAlgo = HashAlgorithmName.SHA512;
+                curve = ECCurve.NamedCurves.nistP521;
+            }
+            else
+            {
+                error = new FormatException($"ECDSA curve '{curveName}' is unsupported.");
+                return false;
+            }
+
+            ECPoint q = reader.ReadStringAsECPoint();
+            BigInteger d = reader.ReadMPInt();
+
+            ECDsa ecdsa = ECDsa.Create();
+            try
+            {
+                byte[] dBytes = d.ToByteArray(true, true);
+                if (dBytes.Length < q.X!.Length)
                 {
-                    rsa.ImportParameters(parameters);
-                    privateKey = new RsaPrivateKey(rsa);
-                    error = null;
-                    return true;
+                    Array.Resize(ref dBytes, q.X.Length);
                 }
-                catch (Exception ex)
+
+                ECParameters parameters = new()
                 {
-                    error = new FormatException($"The data can not be parsed into an RSA key.", ex);
-                    rsa?.Dispose();
-                    return false;
+                    Curve = curve,
+                    D = d.ToByteArray(true, true),
+                    Q = q,
+                };
+                if (parameters.D.Length < q.X!.Length)
+                {
+                    // ToByteArray will shorten the value to the minimum size,
+                    // we need to prepend 0 to the array to make it the
+                    // size.
+                    byte[] originalD = parameters.D;
+                    byte[] expandedD = new byte[q.X.Length];
+                    Buffer.BlockCopy(originalD, 0, expandedD, expandedD.Length - originalD.Length, originalD.Length);
+                    parameters.D = expandedD;
```

Comment:
I'm not 100% sure on this one. I don't know if this is a side effective of the P521 curve or just the ECDSA cert I generated but it was always 1 byte short. I couldn't just use `ReadStringAsByteArray()` as that also returned the shorter byte array.

Reply:
Can you check if `BigInteger d` is a positive or a negative number.

Reply:
From https://github.com/tmds/Tmds.Ssh/pull/206/files#r1693915325, I assume something expects parameters.D.Length and q.X!.Length to be the same.

Then it makes sense to have to pre-pend D with zeros when it is a big-endian unsigned value.
That way we meet the length expectation without changing the numeric value of D.

Reply:
> I assume something expects parameters.D.Length and q.X!.Length to be the same.

Yep importing the parameters in the `ECDsa` object fails if the length of `D` does not match the curve lengths. When the value is signed it contains an extra byte and I found in the P521 example it didn't contain enough bytes so needed to be left padded with 0s.

> Can you check if BigInteger d is a positive or a negative number.

Using the 3 different curves set up in the test private key we have the following values

P256 - 8.35053129e+76 - Sign 1
00B89E550B526C3BEF46B130997C32560452C5011EFDFFB48820CE38D91F6AD2FD
00B89E550B526C3BEF46B130997C32560452C5011EFDFFB48820CE38D91F6AD2FD
B89E550B526C3BEF46B130997C32560452C5011EFDFFB48820CE38D91F6AD2FD


P384 - 2.45161837e+115 - Sign 1
009F48EC8F7EB1705D63DB6AD3F26A6B37E340E7944F30F7EC2CDCC65337AED6472F95721136013C20659CC4558C0002FF
009F48EC8F7EB1705D63DB6AD3F26A6B37E340E7944F30F7EC2CDCC65337AED6472F95721136013C20659CC4558C0002FF
9F48EC8F7EB1705D63DB6AD3F26A6B37E340E7944F30F7EC2CDCC65337AED6472F95721136013C20659CC4558C0002FF


P521 - 1.00062567e+155 - Sign 1
077687A370EF8BD1C78C891537CA857AFDB32FC95D56CD5CD501F527F399CC17C136D8C8CF0FB630E4C62DEAFF61CCDFC4F6C0272AC92995FB603214DF86CC2EE8
077687A370EF8BD1C78C891537CA857AFDB32FC95D56CD5CD501F527F399CC17C136D8C8CF0FB630E4C62DEAFF61CCDFC4F6C0272AC92995FB603214DF86CC2EE8
077687A370EF8BD1C78C891537CA857AFDB32FC95D56CD5CD501F527F399CC17C136D8C8CF0FB630E4C62DEAFF61CCDFC4F6C0272AC92995FB603214DF86CC2EE8

The first value is the raw MPInt bytes after the length, the second is `ToByteArray` as a signed value and the latter is as a unsigned value.

It looks like I should be doing `ToByteArray(isUnsigned: false, isBigEndian: true)` but I need to handle these two both the extra 0 padding that makes the length too long and the case when there isn't enough padding to begin with.

I'll do some testing and generate some more ECDSA certs to see if there are other scenarios I'm not thinking off here.

Reply:
The latest commits change the logic here to just read the raw bytes and handle when D is the required size, contains the leading 0 byte sign char, needs to be left padded. I've also updated the test keys where P256 is the required size, P384 contains the 0 byte sign char, and P521 needs to be left padded with 0s.

Reply:
I pushed a commit that moved the padding logic into the SequenceReader.